Add new parameter handling scheme for Barotropic model

bgroenks96 opened the pull request:
Resolves #753 (see comments therein)

Note that I still need (?) to go back through and add bounds to the parameters. I am not sure if this is explicitly needed at the moment, but I think it's usually helpful to have.

@milankl and @maximilian-gelbrecht please let me know if you agree with (1) the general scheme, and (2) the choice of which types/fields to mark as parameters. I did this relatively naively so it would be good to get feedback on that.

Remaining tasks
- [x] Add optional parameter argument to `initialize` which automatically updates the model struct in the simulation with the given parameters.
- [x] Add `@parameterized` macro for more user-friendly parameter specification
- [x] Move parameter types and methods into new `Utils` module
- [ ] ~~Update I/O to write parameter table instead of `parameters.txt`~~ (wait until next PR)
- [x] Test compatibility with Enzyme

bgroenks96 committed:
Add new package dependencies

bgroenks96 committed:
Add SpeedyParam and parameter utilities

bgroenks96 committed:
Fix syntax in AbstractModel constructor

bgroenks96 committed:
Initial impl of parameters method for Barotropic model types

bgroenks96 committed:
Generalize parameters type arg and add tests

bgroenks96 committed:
Remove outdated comment

bgroenks96 commented on src/SpeedyWeather.jl:
```diff
@@ -24,6 +27,11 @@ import JLD2: jldopen, jldsave, JLDFile
+import ConstructionBase: constructorof, getproperties, setproperties
+
+# DOMAINS
+using DomainSets
```
@milankl I know you prefer `import` generally but I think `using` is easier for `DomainSets`. Also, VS Code incorrectly highlights `import DomainSets: :(..)` (the interval operator) as a syntax error which is kind of annoying...

milankl replied:
Fine with me

maximilian-gelbrecht reviewed (commented):
Really cool. This looks good so far! 

So, with this, how would the work flow look like if we only want to make one specific of the parameters of a certain struct tunable?

bgroenks96 commented on src/parameters.jl:
```diff
@@ -0,0 +1,90 @@
+    desc::String = ""
+
+    "additional attributes with unspecified types"
+    attrs::NamedTuple
```
I am aware, but I created SpeedyParam to get around needing to add type parameters to the model types. I opted for the abstract type here because it (presumably) will not be used or invoked during model execution since we would have already reconstructed the type before that. I think we would be able to avoid the Enzyme code ever seeing SpeedyParam at all, actually. It's really here just as a user interface feature.

Since I adopted the non-invasive approach here, though, I guess SpeedyParam is no longer strictly necessary since we're not assigning Param types to struct fields directly.

maximilian-gelbrecht replied:
Okay, understood!

bgroenks96 commented:
@maximilian-gelbrecht Related question, is `initialize!` parameter-dependent? i.e. when optimizing/sampling/etc. would you include a call to `initialize!` in each evaluation or pre-initialize and then only call `run!`?

maximilian-gelbrecht commented:
I think the best cause of action would be set up a simple test case with Enzyme to test all of these things. 

We should probably also merge #680 for that. I can update it as soon as I find some time.

bgroenks96 committed:
Remove unnecessary type parameters

bgroenks96 committed:
Add parameterof method and SpeedyParams type

SpeedyParams is an implementation of ModelParameters.AbstractModel that allows the user to interact with the parameters via a table-like interface. It also supports taking subsets of parameters.

bgroenks96 commented:
Ok still need tests but subsetting parameters is now supported via `SpeedyParams`, which is a customized implementation of the `AbstractModel` interface from ModelParameters (confusing, I know...).

## Table interface

```julia
spectral_grid = SpectralGrid(trunc=31, nlayers=1)   # define resolution
model = BarotropicModel(spectral_grid)
model_ps = SpeedyParams(parameters(model))
```

```
SpeedyParams{@NamedTuple{planet::@NamedTuple{rotation::SpeedyParam{Float32}, gravity::SpeedyParam{Float32}, axial_tilt::SpeedyParam{Float32}, solar_constant::SpeedyParam{Float32}}, atmosphere::@NamedTuple{mol_mass_dry_air::SpeedyParam{Float32}, mol_mass_vapour::SpeedyParam{Float32}, heat_capacity::SpeedyParam{Float32}, R_gas::SpeedyParam{Float32}, R_dry::SpeedyParam{Float32}, R_vapour::SpeedyParam{Float32}, mol_ratio::SpeedyParam{Float32}, μ_virt_temp::SpeedyParam{Float32}, κ::SpeedyParam{Float32}, water_density::SpeedyParam{Float32}, latent_heat_condensation::SpeedyParam{Float32}, latent_heat_sublimation::SpeedyParam{Float32}, stefan_boltzmann::SpeedyParam{Float32}, pres_ref::SpeedyParam{Float32}, temp_ref::SpeedyParam{Float32}, moist_lapse_rate::SpeedyParam{Float32}, dry_lapse_rate::SpeedyParam{Float32}, layer_thickness::SpeedyParam{Float32}}, forcing::@NamedTuple{strength::SpeedyParam{Float32}, wavenumber::SpeedyParam{Float32}}, drag::@NamedTuple{c::SpeedyParam{Float32}}}} with parent object of type: 

@NamedTuple{planet::@NamedTuple{rotation::SpeedyParam{Float32}, gravity::SpeedyParam{Float32}, axial_tilt::SpeedyParam{Float32}, solar_constant::SpeedyParam{Float32}}, atmosphere::@NamedTuple{mol_mass_dry_air::SpeedyParam{Float32}, mol_mass_vapour::SpeedyParam{Float32}, heat_capacity::SpeedyParam{Float32}, R_gas::SpeedyParam{Float32}, R_dry::SpeedyParam{Float32}, R_vapour::SpeedyParam{Float32}, mol_ratio::SpeedyParam{Float32}, μ_virt_temp::SpeedyParam{Float32}, κ::SpeedyParam{Float32}, water_density::SpeedyParam{Float32}, latent_heat_condensation::SpeedyParam{Float32}, latent_heat_sublimation::SpeedyParam{Float32}, stefan_boltzmann::SpeedyParam{Float32}, pres_ref::SpeedyParam{Float32}, temp_ref::SpeedyParam{Float32}, moist_lapse_rate::SpeedyParam{Float32}, dry_lapse_rate::SpeedyParam{Float32}, layer_thickness::SpeedyParam{Float32}}, forcing::@NamedTuple{strength::SpeedyParam{Float32}, wavenumber::SpeedyParam{Float32}}, drag::@NamedTuple{c::SpeedyParam{Float32}}}


Parameters:
┌───────┬──────────────────────────┬──────────┬──────────────────────────────┬────────────┬─────────────────────────────────────┬────────────────────────────────────────────────────────────────────┐
│   idx │                fieldname │      val │                componenttype │  component │                              bounds │                                                               desc │
│ Int64 │                   Symbol │  Float32 │                     DataType │     Symbol │ IntervalSets.TypedEndpointsInterval │                                                             String │
├───────┼──────────────────────────┼──────────┼──────────────────────────────┼────────────┼─────────────────────────────────────┼────────────────────────────────────────────────────────────────────┤
│     1 │                 rotation │  7.29e-5 │               Earth{Float32} │     planet │       -Inf .. Inf (open) (RealLine) │                      angular frequency of Earth's rotation [rad/s] │
│     2 │                  gravity │     9.81 │               Earth{Float32} │     planet │ 0.0 .. Inf (closed-open) (HalfLine) │                                 gravitational acceleration [m/s^2] │
│     3 │               axial_tilt │     23.4 │               Earth{Float32} │     planet │                         -180 .. 180 │                          angle [˚] rotation axis tilt wrt to orbit │
│     4 │           solar_constant │   1365.0 │               Earth{Float32} │     planet │ 0.0 .. Inf (closed-open) (HalfLine) │              Total solar irradiance at the distance of 1 AU [W/m²] │
│     5 │         mol_mass_dry_air │  28.9649 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │                                      molar mass of dry air [g/mol] │
│     6 │          mol_mass_vapour │  18.0153 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │                                 molar mass of water vapour [g/mol] │
│     7 │            heat_capacity │   1004.0 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │                     specific heat at constant pressure cₚ [J/K/kg] │
│     8 │                    R_gas │   8.3145 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │                                   universal gas constant [J/K/mol] │
│     9 │                    R_dry │  287.054 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │                         specific gas constant for dry air [J/kg/K] │
│    10 │                 R_vapour │  461.524 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │                    specific gas constant for water vapour [J/kg/K] │
│    11 │                mol_ratio │  0.62197 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │ Ratio of gas constants: dry air / water vapour, often called ε [1] │
│    12 │              μ_virt_temp │ 0.607794 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │                          Virtual temperature Tᵥ calculation factor │
│    13 │                        κ │ 0.285911 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │                   = R_dry/cₚ, gas const for air over heat capacity │
│    14 │            water_density │   1000.0 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │                                              water density [kg/m³] │
│    15 │ latent_heat_condensation │  2.501e6 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │                                 latent heat of condensation [J/kg] │
│    16 │  latent_heat_sublimation │  2.801e6 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │                                  latent heat of sublimation [J/kg] │
│    17 │         stefan_boltzmann │  5.67e-8 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │                                stefan-Boltzmann constant [W/m²/K⁴] │
│    18 │                 pres_ref │ 100000.0 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │                                    surface reference pressure [Pa] │
│    19 │                 temp_ref │    288.0 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │                                  surface reference temperature [K] │
│    20 │         moist_lapse_rate │    0.005 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │             reference moist-adiabatic temperature lapse rate [K/m] │
│    21 │           dry_lapse_rate │   0.0098 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │               reference dry-adiabatic temperature lapse rate [K/m] │
│    22 │          layer_thickness │   8500.0 │     EarthAtmosphere{Float32} │ atmosphere │       -Inf .. Inf (open) (RealLine) │                    layer thickness for the shallow water model [m] │
│    23 │                 strength │  3.0e-12 │      KolmogorovFlow{Float32} │    forcing │       -Inf .. Inf (open) (RealLine) │                                         Strength of forcing [1/s²] │
│    24 │               wavenumber │      8.0 │      KolmogorovFlow{Float32} │    forcing │       -Inf .. Inf (open) (RealLine) │       Wavenumber of forcing in meridional direction (pole to pole) │
│    25 │                        c │   1.0e-7 │ LinearVorticityDrag{Float32} │       drag │       -Inf .. Inf (open) (RealLine) │                            Linear vorticity drag coefficient [1/s] │
└───────┴──────────────────────────┴──────────┴──────────────────────────────┴────────────┴─────────────────────────────────────┴────────────────────────────────────────────────────────────────────┘

```

One quick note about the table interface: `ModelParameters` doesn't fully implement tabular behavior; e.g. it only supports column indexing and not row indexing, which would in principle also allow for flexibly selecting subsets of parameters based on their attributes. I had a [PR](https://github.com/rafaqz/ModelParameters.jl/pull/44) for this a long time ago but it never got merged. But I think @rafaqz is open to putting some development effort into ModelParameters again so that's something I can consider taking another look at.

## Parameter subsets

For now, parameter subsets can be extracted with the `ComponentArrays` label syntax:

```julia
model_ps_subset = model_ps[["planet.solar_constant","atmosphere.heat_capacity"]]
```

```
Parameters:
┌───────┬────────────────┬─────────┬──────────────────────────┬────────────┬─────────────────────────────────────────────────────┬───────────────────────────────────────────────────────┐
│   idx │      fieldname │     val │            componenttype │  component │                                              bounds │                                                  desc │
│ Int64 │         Symbol │ Float32 │                 DataType │     Symbol │ DomainSets.FixedInterval{L, :open, Float64} where L │                                                String │
├───────┼────────────────┼─────────┼──────────────────────────┼────────────┼─────────────────────────────────────────────────────┼───────────────────────────────────────────────────────┤
│     1 │ solar_constant │  1365.0 │           Earth{Float32} │     planet │                 0.0 .. Inf (closed-open) (HalfLine) │ Total solar irradiance at the distance of 1 AU [W/m²] │
│     2 │  heat_capacity │  1004.0 │ EarthAtmosphere{Float32} │ atmosphere │                       -Inf .. Inf (open) (RealLine) │        specific heat at constant pressure cₚ [J/K/kg] │
└───────┴────────────────┴─────────┴──────────────────────────┴────────────┴─────────────────────────────────────────────────────┴───────────────────────────────────────────────────────┘
```

A `CompoentVector` can be constructed from SpeedyParams with `vec`:

```julia
pvec = vec(model_ps)
```

```
ComponentVector{Float32}(planet = (rotation = 7.29f-5, gravity = 9.81f0, axial_tilt = 23.4f0, solar_constant = 1365.0f0), atmosphere = (mol_mass_dry_air = 28.9649f0, mol_mass_vapour = 18.0153f0, heat_capacity = 1004.0f0, R_gas = 8.3145f0, R_dry = 287.05432f0, R_vapour = 461.52438f0, mol_ratio = 0.62197006f0, μ_virt_temp = 0.60779446f0, κ = 0.2859107f0, water_density = 1000.0f0, latent_heat_condensation = 2.501f6, latent_heat_sublimation = 2.801f6, stefan_boltzmann = 5.67f-8, pres_ref = 100000.0f0, temp_ref = 288.0f0, moist_lapse_rate = 0.005f0, dry_lapse_rate = 0.0098f0, layer_thickness = 8500.0f0), forcing = (strength = 3.0f-12, wavenumber = 8.0f0), drag = (c = 1.0f-7))
```

I will add tests and reasonable `bounds` for all parameters tomorrow. Also, if @milankl is up for it, I think it would be nice to add `Unitful` as another dependency (sorry for adding so many...) and have first-class units as a column in the parameter table.

milankl commented:
1. We currently write `parameters.txt` with output, `CSV.write("parameters.csv", model_ps)` looks like a neater approach feel free to use that instead (in another PR probably though)

2. > Also, if @milankl is up for it, I think it would be nice to add Unitful as another dependency (sorry for adding so many...) and have first-class units as a column in the parameter table.

I'm not strictly against Unitful just careful. So question can we first keep the units in the docstrings/description and then if everything else works get the units in? Unitful seems to be relatively light weight so don't see big problems adding this as a dependency

bgroenks96 commented on src/models/barotropic.jl:
```diff
@@ -61,6 +61,13 @@ end
 prognostic_variables(::Type{<:Barotropic}) = (:vor,)
 default_concrete_model(::Type{Barotropic}) = BarotropicModel
 
+parameters(model::Barotropic; kwargs...) = (
```
Yes, they can always be stacked, as many times as you want :)

bgroenks96 commented on src/parameters.jl:
```diff
@@ -0,0 +1,196 @@
+const AbstractParam = ModelParameters.AbstractParam
```
Could also do that, but then you have to write `import ModelParameters: ModelParameters, AbstractParam` which always kinds of annoys me.

Then again, I guess it's not worse than this...

bgroenks96 committed:
Add @parameterized macro for parameteric structs

bgroenks96 committed:
Add missing import and fix macro bugs

bgroenks96 committed:
Wrap parameters in SpeedyParams by default

bgroenks96 commented:
@milankl Since I value your happiness, I have attempted to implement your ideal solution with a macro `@parameterized`.

Here is an example for `Earth`:

```julia
@parameterized @kwdef mutable struct Earth{NF<:AbstractFloat} <: AbstractPlanet

    "angular frequency of Earth's rotation [rad/s]"
    @param rotation::NF = DEFAULT_ROTATION
    
    "gravitational acceleration [m/s^2]"
    @param gravity::NF = DEFAULT_GRAVITY (bounds=Nonnegative,)
    
    "switch on/off daily cycle"
    daily_cycle::Bool = true
    
    "Seconds in a daily rotation"
    length_of_day::Second = Hour(24)

    "switch on/off seasonal cycle"
    seasonal_cycle::Bool = true

    "Seconds in an orbit around the sun"
    length_of_year::Second = Day(365.25)
    
    "time of spring equinox (year irrelevant)"
    equinox::DateTime = DateTime(2000, 3, 20) 

    "angle [˚] rotation axis tilt wrt to orbit"
    @param axial_tilt::NF = 23.4 (bounds=-90..90,)

    "Total solar irradiance at the distance of 1 AU [W/m²]"
    @param solar_constant::NF = 1365 (bounds=Nonnegative,)
end
```
The `@param` is not a real macro just a marker... which is maybe confusing. I have mixed feelings about this since I generally don't love introducing non-standard Julia syntax... but maybe the win here is big enough to merit it.

```julia
model_ps = parameters(Earth(Float16))

Parameters:
┌───────┬────────────────┬─────────┬────────────────┬─────────────────────────────────────┬───────────────────────────────────────────────────────┐
│   idx │      fieldname │     val │  componenttype │                              bounds │                                                  desc │
│ Int64 │         Symbol │ Float16 │       DataType │ IntervalSets.TypedEndpointsInterval │                                                String │
├───────┼────────────────┼─────────┼────────────────┼─────────────────────────────────────┼───────────────────────────────────────────────────────┤
│     1 │       rotation │ 7.29e-5 │ Earth{Float16} │       -Inf .. Inf (open) (RealLine) │         angular frequency of Earth's rotation [rad/s] │
│     2 │        gravity │    9.81 │ Earth{Float16} │ 0.0 .. Inf (closed-open) (HalfLine) │                    gravitational acceleration [m/s^2] │
│     3 │     axial_tilt │    23.4 │ Earth{Float16} │                           -90 .. 90 │             angle [˚] rotation axis tilt wrt to orbit │
│     4 │ solar_constant │ 1.365e3 │ Earth{Float16} │ 0.0 .. Inf (closed-open) (HalfLine) │ Total solar irradiance at the distance of 1 AU [W/m²] │
└───────┴────────────────┴─────────┴────────────────┴─────────────────────────────────────┴───────────────────────────────────────────────────────┘
```

What do you think?

maximilian-gelbrecht commented:
I get your concerns about non-Julia syntax and too many macros, but this looks really almost exactly like what we want to have. 

If it works reliably and doesn't hurt performance, I would be in favour of it. 

Btw: I think we should probably put these parameter related functions into another new submodule, or in the `Utils` submodule once #727 is merged

milankl commented on src/parameters.jl:
```diff
@@ -172,6 +174,82 @@ function reconstruct(obj, values::NamedTuple{keys}) where {keys}
+        # case 4: field documentation
+        elseif @capture(ex, docs_String)
+            # store in lastdoc field
+            lastdoc = docs
```
This technically would allow

```julia
    "doc string"
    # empty line here
    a::T
```

no? Because I think DocStringExtensions.jl doesn't allow that?!

bgroenks96 replied:
Yes, it would allow that, but I don't think you should do it?

milankl replied:
Yeah no. Just thought out loud here. I think it's fine don't see this causing problems

milankl reviewed (commented):
This is awesome! And I agree with Max likely exactly what we want!

milankl commented:
> If it works reliably and doesn't hurt performance, I would be in favour of it.

If I understand this correctly `parameters(::Earth)` is now defined *every time* one constructs an `earth = Earth(Float16)` right? But the `earth` is then exactly the same as without `@parameterized` so there cannot be any performance penalty?

milankl commented:
Somewhat unrelated but I struggled with this for a while
```julia
julia> Earth()
ERROR: MethodError: no method matching Earth(::Float64, ::Float64, ::Bool, ::Hour, ::Bool, ::Second, ::DateTime, ::Float64, ::Int64)
The type `Earth` exists, but no method is defined for this combination of argument types when trying to construct it.
```

while the constructor `Earth()` is defined automatically through `@kwdef` (and therefore cannot be redefined manually without causing precompilation issues) it can't do the type conversion Hour -> Second and Int64 -> Float64 without the type parameter `NF` in `Earth{NF}` being specified. That's why we have `Earth(::Type{NF})` and `Earth(::SpectralGrid)` methods. But I feel it counterintuitive that one cannot just create an `Earth()` ... 😠

bgroenks96 commented:
> If I understand this correctly parameters(::Earth) is now defined every time one constructs an earth = Earth(Float16) right?

No, it gets defined once when the macro is expanded prior to compilation, which should happen when you do `using SpeedyWeather`.

But yes, it really cannot possibly have any effect on performance. That was the point of making it non-invasive!

bgroenks96 commented:
> the constructor Earth() is defined automatically through @kwdef (and therefore cannot be redefined manually without causing precompilation issues) it can't do the type conversion Hour -> Second and Int64 -> Float64 without the type parameter NF in Earth{NF} being specified. That's why we have Earth(::Type{NF}) and Earth(::SpectralGrid) methods. But I feel it counterintuitive that one cannot just create an Earth() ... 😠

Why can't it be redefined...? It's not obvious to me why that would cause precompilation issues.

bgroenks96 commented on src/parameters.jl:
```diff
@@ -125,6 +124,9 @@ Base.getindex(ps::SpeedyParams, nm::Symbol) = getindex(ps, :, nm)
 
+## non-mutating setindex!
+## consider using a method instead of setindex! to avoid confusion?
+Base.setindex!(ps::SpeedyParams, x, nm::Symbol) = setindex!(ps, x, :, nm)
```
Yes this is just because of `ModelParameters`. I would like to change it, and we probably will change it soon. But I think this needs to stay the way it is for now in order to be consistent with the `Model` interface.

bgroenks96 commented on src/parameters.jl:
```diff
@@ -11,7 +9,7 @@ Base.@kwdef struct SpeedyParam{NF<:AbstractFloat} <: AbstractParam{NF}
     
     "numerical domain on which the parameter is defined"
-    bounds::Domain = RealLine()
+    bounds::Domain = ℝ
```
I thought I removed this... but yes I will change it back.

milankl commented:
Oh you can't redefine a method otherwise you get precompilation warnings! Try it 😉

milankl commented on src/SpeedyWeather.jl:
```diff
@@ -27,13 +27,17 @@ import JLD2: jldopen, jldsave, JLDFile
+import DomainSets: Domain, RealLine, NonnegativeRealLine, PositiveRealLine, NegativeRealLine, UnitInterval
 using DomainSets.IntervalSets
 
+const StrictlyPositive = PositiveRealLine()
+const Nonnegative = NonnegativeRealLine()
+const Negative = NegativeRealLine()
```
Sounds good!! Even though `Positive` sounds ambiguous somehow out of context it's not clear it refers to bounds?

bgroenks96 replied:
Hmm... it doesn't necessarily have to refer to bounds actually. You could use it in other contexts. It's just a domain.

bgroenks96 committed:
Rename constant StrictlyPositive to Positive

bgroenks96 committed:
More improvements to parameter handling interface

- Make @parameterized more robust
- Add setproperties dispatch to @parameterized for kwdef'd structs
-  Convert other Barotropic model types to @parameterized interface

bgroenks96 committed:
Add dispatch for initialize!

bgroenks96 committed:
Make parameters construction type stable

bgroenks96 commented:
@milankl Please review the choices of  `bounds` on all parameters!

bgroenks96 committed:
Update changelog

bgroenks96 committed:
Add missing dependencies to test project

bgroenks96 committed:
Temporary workaround for failing test

bgroenks96 committed:
Fix namespace conflict in parameters tests

bgroenks96 committed:
Move parameter utils to Utils module

bgroenks96 commented on src/Utils/parameters.jl:
```diff
@@ -0,0 +1,343 @@
+"""
+    $SIGNATURES
+
+Returns a "parameterized" copy of the given data structure. For numeric types, this simply wraps
```
TODO: fix this out-of-date docstring

milankl committed:
bounds updated

milankl reviewed (commented):
Changed some bounds with the last commit!

maximilian-gelbrecht commented on src/Utils/parameters.jl:
```diff
@@ -0,0 +1,343 @@
+reconstruct(obj::T, value::T) where {T} = value
+reconstruct(obj::AbstractParam, value::T) where {T} = ModelParameters.update(obj, Tuple(value))
+reconstruct(obj::NamedTuple{keys,V}, values::NamedTuple{keys,V}) where {keys,V<:Tuple} = values
+reconstruct(obj, values::ComponentArray) = reconstruct(obj, NamedTuple(values))
```
Just wondering, if we generally keep the parameters in a ComponentArray in ML tasks, maybe it would be better to have a routine that directly reconstructs from a ComponentArray as well. Or is `NamedTuple(::ComponentArray)` trivial and non-allocating?

maximilian-gelbrecht reviewed (commented):
I am not so familiar with DomainSets. What happens when trying to set a parameter with an out of bounds value? Just a straight out of bounds error? I am even wondering what kind of behaviour we would desire when using this in optimisation tasks. Maybe I'd even prefer it staying at the lower/upper bound and just raising a warning.

bgroenks96 committed:
Remove @param on R_dry

bgroenks96 committed:
Fix dispatch bug with parameters method

bgroenks96 committed:
Fix type stability of reconstruct for nested structs

bgroenks96 committed:
Add simple parameter differentiability tests for Barotropic model

milankl commented on test/parameters.jl:
```diff
@@ -0,0 +1,123 @@
+using ComponentArrays: ComponentVector
+using DomainSets: Domain, RealLine, UnitInterval
+using SpeedyWeather: value, bounds, description, attributes, parameters, reconstruct, stripparams
+
+import ModelParameters: ModelParameters, Model, Param, params, update
```
`parameters`, `stripparams`, `Param`, `params` are somewhat inconsistetly named. Happy to keep the current if there's a good reason for it but maybe there isn't?

milankl reviewed (commented):
Looks great to me! I'd be happy to merge this without putting too much additional thought or effort into the design. It's already waaaay beyond an initial draft of how parameter handling could be done and seems to have some fantastic potential! I believe the first use cases of this will more easily reveal how to adapt the parameter handling and then this can be updated in another PR!

bgroenks96 committed:
Refactor Barotoropic differentiability tests

bgroenks96 commented:
> Maybe one last thing: Documentation of the parameter handling would come in another pull request when it's actually implemented for all models?

Yes.

> Could be as short as "this is how to retrieve all parameters of a model, how to change them and reconstruct a new model from those" and then "more to come"?

Sure, but do you think we should export `reconstruct`? I would rather not because it conflicts with Flatten.jl.

I'll try to think of another name...

bgroenks96 committed:
Update docs on parameter handling

bgroenks96 commented:
> How about reparam or withparams?

How about we worry about this in the next PR and just merge this now?

milankl commented:
@torsten20 maybe relevant for what you want ☝🏼

bgroenks96 commented:
> I am in favour of agreeing on the names of the routines now

It's fine, just a find-replace! The main open question is whether or not `reconstruct` (or whatever we call it) should actually be a user-facing API or just something we do internally.
